package/network/config/ltq-vdsl-app/files/10_atm_ptm.sh

```diff
@@ -0,0 +1,48 @@
+#!/bin/sh
```

Comment:
I don't really like the idea of restarting the system without user interaction. Furthermore, I would recommend to use two two hotplug scripts. One for ATM and one for PTM.

Since it is a rare case that the line mode changes, I would prefer to drop the module unload/restart for now and use something like this instead:

```
#!/bin/sh

[ "$DSL_NOTIFICATION_TYPE" = "DSL_STATUS" ] && \
[ "$DSL_TC_LAYER_STATUS" = "ATM" ] && \
! grep -q "ltq_atm_vr9" /proc/modules && exit 0

logger -p daemon.notice -t "dsl-notify" "Switching to TC-Layer ATM"

modprobe ltq_atm_vr9
esi $(printf '%012X' $((1+0x$(tr -d : </sys/class/net/eth0/address))))
```

```
#!/bin/sh

[ "$DSL_NOTIFICATION_TYPE" = "DSL_STATUS" ] && \
[ "$DSL_TC_LAYER_STATUS" = "EFM" ] && \
! grep -q "ltq_ptm_vr9" /proc/modules && exit 0

logger -p daemon.notice -t "dsl-notify" "Switching to TC-Layer EFM/PTM"

modprobe ltq_ptm_vr9
```

I'm not yet fully satisfied with the script. For instance, bringing up the atm ethernet bridge (via init script) isn't necessary if ETM/PTM is used. As far as I know, the whole atm/atm ethernet bridge stuff is exclusivly added for lantiq. I'm not aware of a nother target which supports hardware that make use of ATM.

But I don't have yet an idea how to integrate it nicely without breaking danube/ar9. Do you know by any chance how it is done that the ltq-atm module is autoloaded on danube/ar9 but not on vr9? I fail to see it.

One more concern I have is a race condition. Since ee0091ebed71254d38eef701aebd01bace5a4c1e & 45b6eec9fc4e7262865708f5cdd74dc7502b4fda the carrier flag is set at the time showtime is reached. I would expect it to fail if the netdev doesn't exists at the time showtime is reached. I gave it a quick try on a ptm line and it still seams to work (otherwise netifd wouldn't bring up ppp in my test config). But I'm not sure if it is a lucky coincident or your changes/the linux netdev subsystem takes care of such situations.

FYI due to the carrier flag change, I'm somehow sure that the esi call isn't required any longer and netifd should handle mac addresses for atm bridges perfect fine. But I wasn't able to test it. As I wrote before, the `esi` call was added to workaround a race condition in a dhcp over atm-bridge configuration, where the first dhcp package was send before the configured mac address was applied to the interface. I can only guess, but due to the move of the `esi` call we might reintroduce this race condition.